### PR TITLE
Dummy Circuit - Basic ECALL

### PR DESCRIPTION
**Pull request by naure**
_Issue #359 and #567_

* The `DummyInstruction` implements all the communications of a step: state, fetch, registers, memory. But it does not verify calculations: any value can be written out.

* Placeholder circuits for missing implementations, including unknown ecalls.

* More precise register op assignment, see #570.

### CODE REVIEW

**Comment by naure** [comment]
Updated + now with ecall support.

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    imm: WitIn,
+}
+
+impl<E: ExtensionField> DummyConfig<E> {
+    fn construct_circuit(
+        circuit_builder: &mut CircuitBuilder<E>,
+        codes: InsnCodes,
+    ) -> Result<Self, ZKVMError> {
+        let (with_rs1, with_rs2, with_rd) = match (codes.format, codes.kind) {
+            (_, InsnKind::EANY) => (true, true, false),
```
Comment (kunxian-xia) [inline ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:71]: For `ecall` instructions, I think they need to read the `rd` register to get `syscall_id`.

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+            (InsnFormat::R, _) => (true, true, true),
+            (InsnFormat::I, _) => (true, false, true),
+            (InsnFormat::S, _) => (true, true, false),
+            (InsnFormat::B, _) => (true, true, false),
+            (InsnFormat::U, _) => (false, false, true),
+            (InsnFormat::J, _) => (false, false, true),
```
Comment (kunxian-xia) [inline ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:77]: Maybe we can get rid of them as opcodes that are not supported (i.e. `DIV / REM / REMU`) right now belong to the `R` category.

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs (hunk 1 above)
Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:71]: It has to match `VMState::ecall`. That’s not true in this PR, I fixed it in another branch. I can backport it here to be consistent.

Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs]: Better idea: just treat it as an I instruction. I’ll make it so in the guest-example. Done.

**Comment by hero78119** [comment]
DummyCircuit implementation LGTM!
Do you think we better make a feature toggle e.g. "dummy" and make it as default feature now?
Via a feature toggling means, at some point, we can assure there is no dummy opcode by toggle off, so in case there are some leak but we dont find it.

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs (hunk 1 above)
Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs]: After reworking this a bit: the ignored ecalls just read the syscall_id and do nothing else. Added comments to explain.

**Comment by naure** [comment]
@hero78119 This is a gadget that an example or a benchmark can use. If you don’t use it, it’s not there.

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs (hunk 1 above)
Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs]: I also turned it into a parameter: `Platform.unsafe_ecall_nop`

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs (hunk 2 above)
Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:79]: Yes we have implemented almost everything in the mean time 😄 . Maybe there will be other use-cases: test, benchmarks, demos ?

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        };
+        let with_mem_write = matches!(codes.category, InsnCategory::Store);
+        let with_mem_read = matches!(codes.category, InsnCategory::Load);
+        let branching = matches!(codes.category, InsnCategory::Branch)
```
[the remark this answers is not in the document]
Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs]: > let the caller decide

Done.

> several memory words

(discussed offline) This is the next step to do in follow-up PRs.

[the remark this answers is not in the document]
Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs]: Tracked in #567.

**Review by kunxian-xia — APPROVED** [review]
Looks good to me! Let's merge it to unblock the integration test #368. Any remaining issue can be addressed in follow-up PRs.

**Pull request merged**